File: trainer.py
"""Tabular Q-learning against a user-supplied reward function, plus greedy
rollout recording for the frontend animation."""

import math
import random

from envs import ENVS, TRAIN_PARAMS


class RewardError(Exception):
    pass


def compile_reward(code):
    """Compile user/Claude-written reward code into a callable.

    The code must define `def reward(prev, action, state): ...` where prev and
    state are the env's state dicts and action is 0-3.
    """
    ns = {"math": math, "abs": abs, "min": min, "max": max, "__builtins__": {
        "abs": abs, "min": min, "max": max, "round": round, "int": int,
        "float": float, "bool": bool, "len": len, "sum": sum, "True": True,
        "False": False, "None": None,
    }}
    try:
        exec(code, ns)
    except Exception as e:
        raise RewardError(f"reward code failed to compile: {e}")
    fn = ns.get("reward")
    if not callable(fn):
        raise RewardError("code must define a function `reward(prev, action, state)`")
    return fn


def validate_reward(env_id, fn):
    env = ENVS[env_id]()
    env.reset()
    prev = env.state_dict()
    env.step(3)
    try:
        r = fn(prev, 3, env.state_dict())
        float(r)
    except Exception as e:
        raise RewardError(f"reward function crashed on a sample step: {e}")


def train(env_id, reward_fn, progress=None, seed=0):
    """Q-learning. Returns the Q-table. Calls progress(ep, total) periodically."""
    p = TRAIN_PARAMS[env_id]
    episodes, alpha, gamma = p["episodes"], p["alpha"], p["gamma"]
    rng = random.Random(seed)
    env = ENVS[env_id]()
    Q = {}

    def q(s):
        if s not in Q:
            Q[s] = [0.0, 0.0, 0.0, 0.0]
        return Q[s]

    decay_until = int(episodes * 0.7)
    for ep in range(episodes):
        eps = max(0.05, 1.0 - ep / decay_until)
        a_lr = max(0.05, alpha * (1.0 - 0.8 * ep / episodes))
        env.reset()
        s = env.obs()
        prev_dict = env.state_dict()
        while not env.done:
            qs = q(s)
            if rng.random() < eps:
                a = rng.randrange(4)
            else:
                m = max(qs)
                a = qs.index(m)
            env.step(a)
            cur_dict = env.state_dict()
            try:
                r = float(reward_fn(prev_dict, a, cur_dict))
            except Exception as e:
                raise RewardError(f"reward function crashed mid-training: {e}")
            ns = env.obs()
            # bootstrap through time-limit truncation; only true terminals stop
            target = r if env.terminal else r + gamma * max(q(ns))
            qs[a] += a_lr * (target - qs[a])
            s, prev_dict = ns, cur_dict
        if progress and ep % 500 == 0:
            progress(ep, episodes)
    if progress:
        progress(episodes, episodes)
    return Q


def rollout(env_id, Q, reward_fn, shift=False):
    """Greedy rollout; returns a list of frames for the frontend."""
    env = ENVS[env_id]()
    env.reset(shift=shift)
    prev_dict = env.state_dict()
    total = 0.0
    frames = [{"step": 0, **env.frame(), "reward": 0.0, "total": 0.0, "events": [],
               "state": prev_dict}]
    while not env.done:
        s = env.obs()
        qs = Q.get(s)
        if qs is None or max(qs) == min(qs) == 0.0:
            # unseen / indifferent state: act greedily anyway (ties -> action 0
            # would look frozen; nudge with a fixed preference order R,D,U,L)
            qs = qs or [0.0] * 4
            order = [3, 1, 0, 2]
            a = max(order, key=lambda i: (qs[i], -order.index(i)))
        else:
            a = qs.index(max(qs))
        events = env.step(a) or []
        cur_dict = env.state_dict()
        try:
            r = float(reward_fn(prev_dict, a, cur_dict))
        except Exception:
            r = 0.0
        total += r
        frames.append({"step": env.steps, **env.frame(), "reward": round(r, 2),
                       "total": round(total, 2), "events": events,
                       "state": cur_dict})
        prev_dict = cur_dict
    return frames

File: trainer.js
/* Tabular Q-learning against a user-supplied reward function, plus greedy
 * rollout recording for the frontend animation. Ported from trainer.py and
 * run entirely in the browser (or in Node, for tests).
 */
(function (root) {
  const ENVS = (typeof require !== "undefined")
    ? require("./envs.js")
    : root.REWARDLAB_ENVS;
  const { CLASSES, TRAIN_PARAMS } = ENVS;

  class RewardError extends Error {}

  // Deterministic PRNG so a given spec trains the same way every time.
  function mulberry32(a) {
    return function () {
      a |= 0; a = (a + 0x6d2b79f5) | 0;
      let t = Math.imul(a ^ (a >>> 15), 1 | a);
      t = (t + Math.imul(t ^ (t >>> 7), 61 | t)) ^ t;
      return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
    };
  }

  function argmax(qs) {
    let bi = 0, bv = qs[0];
    for (let i = 1; i < qs.length; i++) if (qs[i] > bv) { bv = qs[i]; bi = i; }
    return bi;
  }

  /* Compile an LLM/user-written reward function into a callable. The code must
   * define `function reward(prev, action, state) { ... }` returning a number.
   * This runs in the user's own browser tab with their own key — there is no
   * server to attack — but it is still arbitrary JS, so only paste specs you
   * trust to be turned into reward code. */
  function compileReward(code) {
    let fn;
    try {
      fn = new Function(code + "\n;return typeof reward === 'function' ? reward : undefined;")();
    } catch (e) {
      throw new RewardError("reward code failed to compile: " + e.message);
    }
    if (typeof fn !== "function")
      throw new RewardError("code must define a function `reward(prev, action, state)`");
    return fn;
  }

  function validateReward(envId, fn) {
    const env = new CLASSES[envId]();
    env.reset();
    const prev = env.stateDict();
    env.step(3);
    let r;
    try { r = Number(fn(prev, 3, env.stateDict())); }
    catch (e) { throw new RewardError("reward function crashed on a sample step: " + e.message); }
    if (!Number.isFinite(r))
      throw new RewardError("reward function must return a finite number, got " + r);
  }

  // Yield a macrotask so the page can repaint. MessageChannel is NOT throttled
  // in backgrounded/occluded tabs the way setTimeout(0) is (clamped to ~1s),
  // so training stays fast even when the tab isn't focused. Falls back to
  // setTimeout where MessageChannel is unavailable (e.g. older Node).
  const sleep = (typeof MessageChannel !== "undefined")
    ? () => new Promise((res) => {
        const ch = new MessageChannel();
        ch.port1.onmessage = () => { ch.port1.close(); res(); };
        ch.port2.postMessage(0);
      })
    : () => new Promise((res) => setTimeout(res, 0));

  /* Q-learning. Returns the Q-table (Map<obsKey, [4]>). Awaits a yield to the
   * event loop every `yieldEvery` episodes so the page stays responsive and
   * `progress(ep, total)` can repaint. */
  async function train(envId, rewardFn, opts = {}) {
    const { progress = null, seed = 0, yieldEvery = 500 } = opts;
    const p = TRAIN_PARAMS[envId];
    const { episodes, alpha, gamma } = p;
    const rng = mulberry32(seed);
    const env = new CLASSES[envId]();
    const Q = new Map();
    const q = (s) => { let v = Q.get(s); if (!v) { v = [0, 0, 0, 0]; Q.set(s, v); } return v; };

    const decayUntil = Math.floor(episodes * 0.7);
    for (let ep = 0; ep < episodes; ep++) {
      const eps = Math.max(0.05, 1.0 - ep / decayUntil);
      const aLr = Math.max(0.05, alpha * (1.0 - 0.8 * ep / episodes));
      env.reset();
      let s = env.obsKey();
      let prevDict = env.stateDict();
      while (!env.done) {
        const qs = q(s);
        const a = rng() < eps ? Math.floor(rng() * 4) : argmax(qs);
        env.step(a);
        const curDict = env.stateDict();
        let r;
        try { r = Number(rewardFn(prevDict, a, curDict)); }
        catch (e) { throw new RewardError("reward function crashed mid-training: " + e.message); }
        if (!Number.isFinite(r)) r = 0;
        const ns = env.obsKey();
        // bootstrap through time-limit truncation; only true terminals stop
        const target = env.terminal ? r : r + gamma * Math.max.apply(null, q(ns));
        qs[a] += aLr * (target - qs[a]);
        s = ns; prevDict = curDict;
      }
      if (progress && ep % 500 === 0) progress(ep, episodes);
      if (ep % yieldEvery === 0) await sleep();
    }
    if (progress) progress(episodes, episodes);
    return Q;
  }

  const round2 = (x) => Math.round(x * 100) / 100;

  /* Greedy rollout; returns a list of frames for the frontend. */
  function rollout(envId, Q, rewardFn, opts = {}) {
    const { shift = false, seed = 0 } = opts;
    const rng = mulberry32(seed + (shift ? 1 : 0));
    const env = new CLASSES[envId]();
    env.reset(shift, rng);
    let prevDict = env.stateDict();
    let total = 0;
    const frames = [Object.assign(
      { step: 0, reward: 0, total: 0, events: [], state: prevDict }, env.frame())];
    const order = [3, 1, 0, 2]; // preference when a state is unseen/indifferent
    while (!env.done) {
      const s = env.obsKey();
      let qs = Q.get(s);
      let a;
      if (!qs || (Math.max.apply(null, qs) === 0 && Math.min.apply(null, qs) === 0)) {
        const q4 = qs || [0, 0, 0, 0];
        a = order.reduce((best, i) =>
          (q4[i] > q4[best] || (q4[i] === q4[best] && order.indexOf(i) < order.indexOf(best))) ? i : best, order[0]);
      } else {
        a = argmax(qs);
      }
      const events = env.step(a) || [];
      const curDict = env.stateDict();
      let r;
      try { r = Number(rewardFn(prevDict, a, curDict)); if (!Number.isFinite(r)) r = 0; }
      catch (e) { r = 0; }
      total += r;
      frames.push(Object.assign(
        { step: env.steps, reward: round2(r), total: round2(total), events, state: curDict },
        env.frame()));
      prevDict = curDict;
    }
    return frames;
  }

  const api = { RewardError, compileReward, validateReward, train, rollout, mulberry32 };
  root.REWARDLAB_TRAINER = api;
  if (typeof module !== "undefined" && module.exports) module.exports = api;
})(typeof window !== "undefined" ? window : globalThis);
